### PR TITLE
[clang][SPIR-V] Always add convergence intrinsics

### LINKED ISSUE #88144: [HLSL] Handle convergence intrinsic generation for do...while & while conditions
Initial convergence intrinsic generation was added through #80680.

The idea was to visit the wave intrinsic, and if one requiring specific convergence control was found, the loop/entry convergence bits were added.
This

### PR DESCRIPTION
**Pull request by Keenuts**
PR #80680 added bits in the codegen to lazily add convergence intrinsics when required. This logic relied on the LoopStack. The issue is when parsing the condition, the loopstack doesn't yet reflect the correct values, as expected since we are not yet in the loop.

However, convergence tokens should sometimes already be available. The solution which seemed the simplest is to greedily generate the tokens when we generate SPIR-V.

Fixes #88144

### CODE REVIEW

File: clang/lib/CodeGen/CGCall.cpp
Comment (arsenm) [inline clang/lib/CodeGen/CGCall.cpp:4829]: can this just be cast<>?

File: clang/lib/CodeGen/CodeGenFunction.h
Comment (arsenm) [inline clang/lib/CodeGen/CodeGenFunction.h:4995]: should just be cast? You also have yet another dyn_cast on the use?

Comment (Keenuts) [reply clang/lib/CodeGen/CodeGenFunction.h:4995]: Forgot to fix the use, and yes, should be a cast. Bad habits...

File: clang/lib/CodeGen/CGCall.cpp
Comment (arsenm) [inline clang/lib/CodeGen/CGCall.cpp:4828]: Shouldn't have this target check

File: clang/lib/CodeGen/CGStmt.cpp
Comment (arsenm) [inline clang/lib/CodeGen/CGStmt.cpp:1127]: None of these target checks have anything to do with the target. If you want to conditionally enable these for bringup, these should be hidden in an enableConvergenceTokenEmission or something predicate

Comment (Keenuts) [reply clang/lib/CodeGen/CGStmt.cpp:1127]: Right, something at the module level, which hides this into a more genering function, like the `shouldEmitRTTI`. Will fix.

Comment (Keenuts) [reply clang/lib/CodeGen/CGStmt.cpp]: Changed those for a module-level function which hides the target specific bit behind a more generic check.
Let me know if that's not what you had in mind!

File: clang/lib/CodeGen/CGStmt.cpp
Comment (bogner) [inline clang/lib/CodeGen/CGStmt.cpp:3168]: Should this be something like `BB->getFirstInsertionPt()`? That's more concise and better behaved around PHI nodes and the like

File: clang/lib/CodeGen/CGStmt.cpp
Comment (bogner) [inline clang/lib/CodeGen/CGStmt.cpp:3181]: LLVM naming prefers `Token`

File: clang/lib/CodeGen/CGStmt.cpp
Comment (llvm-beanz) [inline clang/lib/CodeGen/CGStmt.cpp:3174]: I'm pretty sure that some of this is just code that you moved from CGBuiltin.cpp, but we probably should change the gratuitous use of `auto`. LLVM's conventions only use `auto` when it improves readability (see: https://llvm.org/docs/CodingStandards.html#use-auto-type-deduction-to-make-code-more-readable).

In quite a few of these cases the explicit type would be much more readable.

Comment (Keenuts) [reply clang/lib/CodeGen/CGStmt.cpp]: Right, replaced the auto usage

**Comment by Keenuts** [comment]
Hi all, rebased on main, and addressed the comments.
 This commits changes the register order on SPIR-V vs DXIL, which required me to fix the mad+lerp intrinsic tests. Should be NFC, just storing the register name in a CHECK variable.

File: clang/lib/CodeGen/CodeGenModule.h
Comment (arsenm) [inline clang/lib/CodeGen/CodeGenModule.h:1592]: Should add a TODO this should just be unconditional in the future

**Review by arsenm — APPROVED** [review]

**Review by sudonatalie — APPROVED** [review]

File: clang/lib/CodeGen/CodeGenModule.h
Comment (Keenuts) [reply clang/lib/CodeGen/CodeGenModule.h:1592]: Not necessarily. The SPIR-V compute (as OpenCL not Vulkan compute) doesn't require a structured CFG, so this could remain forever if they don't see any benefit in generating a SCFG.

**Comment by Keenuts** [comment]
Thanks for the reviews. Waiting for 1 approval from MS and I'll merge

**Review by bogner — APPROVED** [review]

File: clang/lib/CodeGen/CodeGenModule.h
Comment (arsenm) [reply clang/lib/CodeGen/CodeGenModule.h:1592]: This doesn't have anything to do with what the target wants to do with with the CFG lowering. Structurally the IR should not allow uncontrolled convergence, and that it exists is a wart that needs to exist until everywhere handles convergence tokens

Comment (Keenuts) [reply clang/lib/CodeGen/CodeGenModule.h:1594]: That makes sense. Added a TODO line!

**Comment by Keenuts** [comment]
rebased on main, local tests are passing, waiting on CI to merge.

**Pull request merged**
